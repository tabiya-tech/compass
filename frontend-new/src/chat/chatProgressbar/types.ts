export enum ConversationPhase {
  INITIALIZING = "INITIALIZING",
  INTRO = "INTRO",
  COLLECT_EXPERIENCES = "COLLECT_EXPERIENCES",
  DIVE_IN = "DIVE_IN",
  ENDED = "ENDED",
  UNKNOWN = "UNKNOWN"
}

export type CurrentPhase = {
  percentage: number
  phase: ConversationPhase
}

export const defaultCurrentPhase = {
  phase: ConversationPhase.INITIALIZING,
  percentage: 0
}
