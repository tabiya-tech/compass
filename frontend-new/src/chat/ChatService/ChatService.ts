import { getServiceErrorFactory } from "src/error/ServiceError/ServiceError";
import { StatusCodes } from "http-status-codes";
import { fetchWithAuth } from "src/utils/fetchWithAuth/fetchWithAuth";
import ErrorConstants from "src/error/ServiceError/ServiceError.constants";
import { getBackendUrl } from "src/envService";
import { ConverstaionResponse } from "./ChatService.types";

export default class ChatService {
  readonly chatEndpointUrl: string;
  readonly chatHistoryEndpointUrl: string;
  readonly apiServerUrl: string;
  private readonly sessionId: number;

  constructor(sessionId: number) {
    this.apiServerUrl = getBackendUrl();
    this.chatEndpointUrl = `${this.apiServerUrl}/conversation`;
    this.chatHistoryEndpointUrl = `${this.apiServerUrl}/conversation/history`;
    this.sessionId = sessionId;
  }

  /**
   * Get the singleton instance of the ChatService.
   * @returns {ChatService} The singleton instance of the ChatService.
   * @param sessionId The session ID to use for the chat service.
   */
  static getInstance(sessionId: number): ChatService {
    return new ChatService(sessionId);
  }

  public getSessionId(): number {
    return this.sessionId;
  }

  public async sendMessage(message: string): Promise<ConverstaionResponse> {
    const serviceName = "ChatService";
    const serviceFunction = "sendMessage";
    const method = "POST";
    const ChatURL = this.chatEndpointUrl;
    const errorFactory = getServiceErrorFactory(serviceName, serviceFunction, method, ChatURL);

    const response = await fetchWithAuth(ChatURL, {
      method: method,
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        session_id: this.getSessionId(),
        user_input: message,
      }),
      expectedStatusCode: StatusCodes.CREATED,
      serviceName,
      serviceFunction,
      failureMessage: `Failed to send message with session id ${this.getSessionId()}`,
      expectedContentType: "application/json",
    });

    const responseBody = await response.text();

    let messageResponse: ConverstaionResponse;
    try {
      messageResponse = JSON.parse(responseBody);
    } catch (e: any) {
      throw errorFactory(
        response.status,
        ErrorConstants.ErrorCodes.INVALID_RESPONSE_BODY,
        "Response did not contain valid JSON",
        {
          responseBody,
          error: e,
        }
      );
    }

    return messageResponse;
  }

  async clearChat(): Promise<void> {
    const serviceName = "ChatService";
    const serviceFunction = "clearChat";
    const method = "GET";
    const qualifiedURL = `${this.chatEndpointUrl}?user_input=&clear_memory=true&session_id=${this.getSessionId()}`;

    await fetchWithAuth(qualifiedURL, {
      method: method,
      headers: { "Content-Type": "application/json" },
      expectedStatusCode: StatusCodes.OK,
      serviceName,
      serviceFunction,
      failureMessage: `Failed to clear chat for session id ${this.getSessionId()}`,
      expectedContentType: "application/json",
    });
  }

  public async getChatHistory(): Promise<ConverstaionResponse> {
    const serviceName = "ChatService";
    const serviceFunction = "getChatHistory";
    const method = "GET";
    const qualifiedURL = `${this.chatHistoryEndpointUrl}?session_id=${this.getSessionId()}`;

    const response = await fetchWithAuth(qualifiedURL, {
      method: method,
      headers: { "Content-Type": "application/json" },
      expectedStatusCode: StatusCodes.OK,
      serviceName,
      serviceFunction,
      failureMessage: `Failed to get chat history for session id ${this.getSessionId()}`,
      expectedContentType: "application/json",
    });

    const responseBody = await response.text();

    let chatHistory: ConverstaionResponse;
    try {
      chatHistory = JSON.parse(responseBody);
    } catch (e: any) {
      const errorFactory = getServiceErrorFactory(serviceName, serviceFunction, method, qualifiedURL);
      throw errorFactory(
        response.status,
        ErrorConstants.ErrorCodes.INVALID_RESPONSE_BODY,
        "Response did not contain valid JSON",
        {
          responseBody,
          error: e,
        }
      );
    }

    return chatHistory;
  }
}
