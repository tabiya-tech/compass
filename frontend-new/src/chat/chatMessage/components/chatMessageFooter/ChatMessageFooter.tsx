import { Typography, styled } from "@mui/material";
import { getDurationFromNow } from "src/utils/getDurationFromNow/getDurationFromNow";

const uniqueId = "7772f20a-9d0c-4072-b24f-97eca2f43d7b";

export const DATA_TEST_ID = {
  CHAT_MESSAGE_TIMESTAMP: `chat-message-sent_at-${uniqueId}`,
};

export interface ChatMessageFooterProps {
  sentAt: string
}

const TimeStamp = styled(Typography)(({ theme }) => ({
  fontSize: theme.typography.body2.fontSize,
  color: theme.palette.text.secondary,
  marginTop: theme.spacing(theme.tabiyaSpacing.xs),
}));


const ChatMessageFooter: React.FC<ChatMessageFooterProps> = ({ sentAt }) => {
  let duration;
  try {
    duration = getDurationFromNow(new Date(sentAt));
  } catch (e) {
    console.error(new Error("Failed to get message duration", { cause: e }));
  }

  return (
    <TimeStamp data-testid={DATA_TEST_ID.CHAT_MESSAGE_TIMESTAMP} variant="caption">
      sent {duration}
    </TimeStamp>
  )
}

export default ChatMessageFooter;