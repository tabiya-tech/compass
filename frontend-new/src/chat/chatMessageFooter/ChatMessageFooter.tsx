import React, { ReactNode } from "react";
import { Box, Divider, useTheme } from "@mui/material";

interface MessageFooterProps {
  children: ReactNode;
}

const uniqueId = "74bf72e8-1ed8-4dcf-8d3a-262d82cdd4b2";

export const DATA_TEST_ID = {
  CHAT_MESSAGE_FOOTER: `chat-message-footer-${uniqueId}`,
  CHAT_MESSAGE_FOOTER_DIVIDER: `chat-message-footer-divider-${uniqueId}`,
};

const ChatMessageFooter: React.FC<MessageFooterProps> = ({ children }) => {
  const theme = useTheme();

  return (
    <Box data-testid={DATA_TEST_ID.CHAT_MESSAGE_FOOTER}>
      <Divider
        color={theme.palette.grey[100]}
        sx={{ marginY: theme.spacing(1) }}
        data-testid={DATA_TEST_ID.CHAT_MESSAGE_FOOTER_DIVIDER}
      />
      {children}
    </Box>
  );
};

export default ChatMessageFooter;
