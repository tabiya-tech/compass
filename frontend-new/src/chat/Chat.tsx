import React, { useCallback, useEffect, useRef, useState } from "react";
import ChatService from "src/chat/ChatService/ChatService";
import ChatList from "src/chat/chatList/ChatList";
import { IChatMessage } from "./Chat.types";
import { generateCompassMessage, generateTypingMessage, generateUserMessage } from "./util";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";
import { Box, useTheme } from "@mui/material";
import ChatHeader from "./ChatHeader/ChatHeader";
import ChatMessageField from "./ChatMessageField/ChatMessageField";
import { getUserFriendlyErrorMessage, RestAPIError } from "src/error/restAPIError/RestAPIError";
import { writeRestAPIErrorToLog } from "src/error/restAPIError/logger";
import { useNavigate } from "react-router-dom";
import { routerPaths } from "src/app/routerPaths";
import UserPreferencesStateService from "src/userPreferences/UserPreferencesStateService";
import { ConversationMessage, ConversationMessageSender } from "./ChatService/ChatService.types";
import { Backdrop } from "src/theme/Backdrop/Backdrop";
import ExperiencesDrawer from "src/experiences/experiencesDrawer/ExperiencesDrawer";
import { Experience } from "src/experiences/experiencesDrawer/experienceService/experiences.types";
import ExperienceService from "src/experiences/experiencesDrawer/experienceService/experienceService";
import UserPreferencesService from "src/userPreferences/UserPreferencesService/userPreferences.service";
import InactiveBackdrop from "src/theme/Backdrop/InactiveBackdrop";
import ConfirmModalDialog from "src/theme/confirmModalDialog/ConfirmModalDialog";
import AuthenticationServiceFactory from "src/auth/services/Authentication.service.factory";
import FeedbackForm from "src/feedback/overallFeedback/feedbackForm/FeedbackForm";
import { ChatError } from "src/error/commonErrors";
import { ChatMessageType } from "src/chat/Chat.types"
import authenticationStateService from "src/auth/services/AuthenticationState.service";

export const INACTIVITY_TIMEOUT = 3 * 60 * 1000; // in milliseconds
// Set the interval to check every TIMEOUT/3,
// so in worst case scenario the inactivity backdrop will show after TIMEOUT + TIMEOUT/3 of inactivity
export const CHECK_INACTIVITY_INTERVAL = INACTIVITY_TIMEOUT + INACTIVITY_TIMEOUT / 3;
const uniqueId = "b7ea1e82-0002-432d-a768-11bdcd186e1d";
export const DATA_TEST_ID = {
  CONTAINER: `container-${uniqueId}`,
  CHAT_CONTAINER: `chat-container-${uniqueId}`,
};

interface ChatProps {
  showInactiveSessionAlert?: boolean;
  disableInactivityCheck?: boolean;
}

const Chat: React.FC<ChatProps> = ({ showInactiveSessionAlert = false, disableInactivityCheck = false }) => {
  const theme = useTheme();
  const { enqueueSnackbar } = useSnackbar();

  const [messages, setMessages] = useState<IChatMessage[]>([]);
  const [conversationCompleted, setConversationCompleted] = useState<boolean>(false);
  const [exploredExperiences, setExploredExperiences] = useState<number>(0);
  const [conversationConductedAt, setConversationConductedAt] = useState<string | null>(null);
  const [currentMessage, setCurrentMessage] = useState<string>("");
  const [isTyping, setIsTyping] = useState<boolean>(false);
  const [isDrawerOpen, setIsDrawerOpen] = useState<boolean>(false);
  const [experiences, setExperiences] = React.useState<Experience[]>([]);
  const [isLoading, setIsLoading] = React.useState<boolean>(false);
  const [isLoggingOut, setIsLoggingOut] = React.useState<boolean>(false);
  const [showBackdrop, setShowBackdrop] = useState(showInactiveSessionAlert);
  const [lastActivityTime, setLastActivityTime] = React.useState<number>(Date.now());
  const [newConversationDialog, setNewConversationDialog] = React.useState<boolean>(false);
  const [exploredExperiencesNotification, setExploredExperiencesNotification] = useState<boolean>(false);
  const [isFeedbackFormOpen, setIsFeedbackFormOpen] = useState<boolean>(false);
  const [activeSessionId, setActiveSessionId] = useState<number | null>(
    UserPreferencesStateService.getInstance().getActiveSessionId()
  );
  const [current_user_id] = useState<string | null>(authenticationStateService.getInstance().getUser()?.id ?? null);
  const [sessionHasFeedback] = useState<boolean>(UserPreferencesStateService.getInstance().activeSessionHasFeedback());

  const navigate = useNavigate();

  const initializationRef = useRef(false);

  /**
   * --- Utility functions ---
   */

  const addMessage = (message: IChatMessage) => {
    setMessages((prevMessages) => [...prevMessages, message]);
  };

  const generateThankYouMessage = () => {
    return generateCompassMessage(
      "Thank you for taking the time to share your valuable feedback. Your input is important to us.",
      new Date().toISOString()
    );
  };

  const checkAndAddConversationConclusionMessage = useCallback(() => {
    // Assumes the conversation is completed
    if (sessionHasFeedback) {
      // If the user has already given feedback, adds a thank-you message
      addMessage(generateThankYouMessage());
    } else {
      // If they haven't, asks the user to give feedback
      addMessage({
        ...generateCompassMessage(
          "We’d love your feedback on this conversation. It’ll only take 5 minutes and will help us improve your experience",
          new Date().toISOString()
        ),
        type: ChatMessageType.CONVERSATION_CONCLUSION,
      });
    }
  }, [sessionHasFeedback]);

  // Depending on the typing state, add or remove the typing message from the messages list
  const addOrRemoveTypingMessage = (userIsTyping: boolean) => {
    if (userIsTyping) {
      // Only add typing message if it doesn't already exist
      setMessages((prevMessages) => {
        // check if the last message is a typing message
        const hasTypingMessage = prevMessages[prevMessages.length - 1]?.type === ChatMessageType.TYPING;

        if (!hasTypingMessage) {
          return [...prevMessages, generateTypingMessage(new Date().toISOString())];
        }
        return prevMessages;
      });
    } else {
      // filter out the typing message
      setMessages((prevMessages) => prevMessages.filter((message) => message.type !== ChatMessageType.TYPING));
    }
  };

  // Issue a new session and update the user preferences
  const issueNewSession = useCallback(
    async (user_id: string | null) => {
      try {
        setMessages([]);
        // If there is no session id, then create a new session
        const preferencesService = UserPreferencesService.getInstance();
        // expect that the user_id is not null, if so then let the getNewSession function handle it
        let user_preferences = await preferencesService.getNewSession(user_id!);
        // Notify the application that the preferences have changed,
        // otherwise when the chat is renderer, the parent component will not have the latest session
        UserPreferencesStateService.getInstance().setUserPreferences(user_preferences);
        enqueueSnackbar("New conversation started", { variant: "success" });
        return UserPreferencesStateService.getInstance().getActiveSessionId()!;
      } catch (e) {
        addMessage(
          generateCompassMessage(
            "I'm sorry, Something seems to have gone wrong on my end... Can you please repeat that?",
            new Date().toISOString()
          )
        );
        enqueueSnackbar("Failed to start new conversation", { variant: "error" });
        console.error(new ChatError("Failed to create new session", e as Error));
      }
      return null;
    },
    [enqueueSnackbar]
  );

  /**
   * --- Service handlers ---
   */

  // Opens the experiences drawer
  // Goes to the experience service to get the experiences
  const handleOpenExperiencesDrawer = useCallback(async () => {
    setIsDrawerOpen(true);
    setIsLoading(true);
    if (!activeSessionId) {
      // If there is no session id, we can't get the experiences
      throw new ChatError("Session id is not available");
    }
    try {
      const experienceService = new ExperienceService();
      const data = await experienceService.getExperiences(activeSessionId);
      setExperiences(data);
      setIsLoading(false);
    } catch (error) {
      enqueueSnackbar("Failed to retrieve experiences", { variant: "error" });
      console.error(new ChatError("Failed to retrieve experiences", error as Error));
    }
  }, [enqueueSnackbar, activeSessionId]);

  // Goes to the authentication service to log the user out
  // Navigates to the login page
  const handleLogout = useCallback(async () => {
    setIsLoggingOut(true);
    const authenticationService = AuthenticationServiceFactory.getCurrentAuthenticationService();
    await authenticationService!.logout();
    navigate(routerPaths.LOGIN, { replace: true });
    enqueueSnackbar("Successfully logged out.", { variant: "success" });
    setIsLoggingOut(false);
  }, [enqueueSnackbar, navigate]);

  // Goes to the chat service to send a message
  const sendMessage = useCallback(
    async (userMessage: string, session_id: number) => {
      setIsTyping(true);
      const sent_at = new Date().toISOString();

      if (currentMessage) {
        // optimistically add the user's message for a more responsive feel
        const message = generateUserMessage(currentMessage, new Date().toISOString());
        addMessage(message);
      }

      try {
        // Send the user's message
        const chatService = ChatService.getInstance(session_id);
        const response = await chatService.sendMessage(userMessage);

        setExploredExperiences(response.experiences_explored);

        if (response.experiences_explored > exploredExperiences) {
          setExploredExperiencesNotification(true);
        }

        response.messages.forEach((messageItem) =>
          addMessage(generateCompassMessage(messageItem.message, messageItem.sent_at))
        );

        setConversationCompleted(response.conversation_completed);
        setConversationConductedAt(response.conversation_conducted_at);

        if (conversationCompleted) {
          checkAndAddConversationConclusionMessage();
        }
      } catch (error) {
        console.error(new ChatError("Failed to send message:", error as Error));
        addMessage(
          generateCompassMessage(
            "I'm sorry, Something seems to have gone wrong on my end... Can you please repeat that?",
            sent_at
          )
        );
      } finally {
        setIsTyping(false);
      }
    },
    [currentMessage, exploredExperiences, conversationCompleted, checkAndAddConversationConclusionMessage]
  );

  const initializeChat = useCallback(
    async (user_id: string | null, currentSessionId: number | null) => {
      setIsTyping(true);
      let sessionId: number | null = currentSessionId;

      try {
        if (!sessionId) {
          sessionId = await issueNewSession(user_id);
          if (!sessionId) {
            console.debug("Failed to issue new session");
            return false;
          }
        }

        // Get the chat history
        const chatService = ChatService.getInstance(sessionId);
        const history = await chatService.getChatHistory();

        // Set the messages from the chat history
        if (history.messages.length) {
          setMessages(
            history.messages.map((message: ConversationMessage) =>
              message.sender === ConversationMessageSender.USER
                ? generateUserMessage(message.message, message.sent_at)
                : generateCompassMessage(message.message, message.sent_at)
            )
          );

          setConversationCompleted(history.conversation_completed);
          setConversationConductedAt(history.conversation_conducted_at);

          // If the conversation is completed, check if the user has given feedback
          if (history.conversation_completed) {
            checkAndAddConversationConclusionMessage();
          }
        } else {
          // if this is the last promise to resolve, we should not set any state before it is resolved
          // This is the first message to kick off the conversation
          await sendMessage("", sessionId);
        }

        // IMPORTANT: set state only after all promises are resolved

        // Set the explored experiences state
        setExploredExperiences(history.experiences_explored);
        setExploredExperiencesNotification(history.experiences_explored > 0);

        // Set the active session id state
        setActiveSessionId(sessionId);
        return true;
      } catch (e) {
        if (e instanceof RestAPIError) {
          writeRestAPIErrorToLog(e, console.error);
        } else {
          console.error(new ChatError("Failed to initialize chat", e as Error));
        }
        const errorMessage = getUserFriendlyErrorMessage(e as Error);
        enqueueSnackbar(errorMessage, { variant: "error" });

        return false;
      } finally {
        setIsTyping(false);
      }
    },
    [issueNewSession, checkAndAddConversationConclusionMessage, sendMessage, enqueueSnackbar]
  );

  // Resets the text field for the next message
  // Optimistically adds the user's message to the messages list
  // Calls the sendMessage function to send the message
  const handleSend = useCallback(async () => {
    if (currentMessage.trim()) {
      setCurrentMessage("");

      await sendMessage(currentMessage, activeSessionId!);
    }
  }, [currentMessage, sendMessage, activeSessionId]);

  /**
   * --- Callbacks for child components ---
   */

  const handleDrawerClose = () => {
    setIsDrawerOpen(false);
  };

  const handleConfirmNewConversation = useCallback(async () => {
    try {
      setNewConversationDialog(false);
      setExploredExperiencesNotification(false);
      await initializeChat(current_user_id, null);
    } catch (e) {
      addMessage(
        generateCompassMessage(
          "I'm sorry, Something seems to have gone wrong on my end... Can you try again?",
          new Date().toISOString()
        )
      );
      enqueueSnackbar("Failed to start new conversation", { variant: "error" });
      console.error(new ChatError("Failed to start new conversation", e as Error));
    }
  }, [enqueueSnackbar, initializeChat, current_user_id]);

  const handleFeedbackSubmit = () => {
    setMessages((prevMessages) =>
      prevMessages.filter((message) => message.type !== ChatMessageType.CONVERSATION_CONCLUSION)
    );
    addMessage(generateThankYouMessage());
  };

  /**
   * --- UseEffects ---
   */

  // Initialize the chat when the component mounts
  useEffect(() => {
    if (initializationRef.current) {
      return;
    }
    initializationRef.current = true;
    initializeChat(current_user_id, activeSessionId).then((initialized: boolean) => {
      initializationRef.current = initialized;
    });
  }, [initializeChat, activeSessionId, current_user_id]);

  // show the user backdrop when the user is inactive for INACTIVITY_TIMEOUT
  useEffect(() => {
    if (disableInactivityCheck || conversationCompleted) return;

    const checkInactivity = () => {
      if (Date.now() - lastActivityTime > INACTIVITY_TIMEOUT) {
        setShowBackdrop(true);
      }
    };
    // Check for inactivity
    const interval = setInterval(checkInactivity, CHECK_INACTIVITY_INTERVAL);

    return () => clearInterval(interval);
  }, [lastActivityTime, disableInactivityCheck, conversationCompleted]);

  // Close backdrop when user interacts with the page
  useEffect(() => {
    if (disableInactivityCheck) return;

    // Reset the timer when the user interacts with the page
    const resetTimer = () => {
      setLastActivityTime(Date.now());
      setShowBackdrop(false);
    };

    const events = ["mousedown", "keydown"];
    events.forEach((event) => document.addEventListener(event, resetTimer));

    return () => events.forEach((event) => document.removeEventListener(event, resetTimer));
  }, [disableInactivityCheck]);

  // add a message when the compass is typing
  useEffect(() => {
    addOrRemoveTypingMessage(isTyping);
  }, [isTyping]);

  return (
    <>
      {isLoggingOut ? (
        <Backdrop isShown={isLoggingOut} message={"Logging you out, wait a moment..."} />
      ) : (
        <>
          <Box
            width="100%"
            height="100%"
            display="flex"
            flexDirection="column"
            position="relative"
            data-testid={DATA_TEST_ID.CHAT_CONTAINER}
          >
            <Box padding={theme.spacing(theme.tabiyaSpacing.xl)}>
              <ChatHeader
                notifyOnLogout={handleLogout}
                startNewConversation={() => setNewConversationDialog(true)}
                notifyOnExperiencesDrawerOpen={handleOpenExperiencesDrawer}
                experiencesExplored={exploredExperiences}
                exploredExperiencesNotification={exploredExperiencesNotification}
                setExploredExperiencesNotification={setExploredExperiencesNotification}
              />
            </Box>
            <Box sx={{ flex: 1, overflowY: "auto", paddingX: theme.tabiyaSpacing.lg }}>
              <ChatList messages={messages} notifyOnFeedbackFormOpened={() => setIsFeedbackFormOpen(true)} />
            </Box>
            {showBackdrop && <InactiveBackdrop isShown={showBackdrop} />}
            <Box sx={{ flexShrink: 0, padding: theme.tabiyaSpacing.lg, paddingTop: theme.tabiyaSpacing.xs }}>
              <ChatMessageField
                handleSend={handleSend}
                aiIsTyping={isTyping}
                isChatFinished={conversationCompleted}
                message={currentMessage}
                notifyChange={setCurrentMessage}
              />
            </Box>
          </Box>
          <ExperiencesDrawer
            isOpen={isDrawerOpen}
            notifyOnClose={handleDrawerClose}
            isLoading={isLoading}
            experiences={experiences}
            conversationConductedAt={conversationConductedAt}
          />
          {newConversationDialog && (
            <ConfirmModalDialog
              isOpen={newConversationDialog}
              title="Start New Conversation?"
              content={
                <>
                  Once you start a new conversation, all messages from the current conversation will be lost forever.
                  <br />
                  <br />
                  Are you sure you want to start a new conversation?
                </>
              }
              onCancel={() => setNewConversationDialog(false)}
              onConfirm={handleConfirmNewConversation}
              cancelButtonText="Cancel"
              confirmButtonText="Yes, I'm sure"
            />
          )}
          <FeedbackForm
            isOpen={isFeedbackFormOpen}
            notifyOnClose={() => setIsFeedbackFormOpen(false)}
            onFeedbackSubmit={handleFeedbackSubmit}
          />
        </>
      )}
    </>
  );
};

export default Chat;
