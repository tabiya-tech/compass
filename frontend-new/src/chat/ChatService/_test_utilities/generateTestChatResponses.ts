import { ConversationMessage, ConversationMessageSender } from "src/chat/ChatService/ChatService.types";

export const generateTestChatResponses = (): ConversationMessage[] => {
  return [
    {
      message: "Hello! How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "I can help you with that. What is your name?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
  ];
};

export const generateTestHistory = (): ConversationMessage[] => {
  return [
    {
      message: "Hello! How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "I can help you with that. What is your name?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
    {
      message: "My name is John.",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.USER,
    },
    {
      message: "Nice to meet you, John. How can I help you today?",
      sent_at: new Date().toISOString(),
      sender: ConversationMessageSender.COMPASS,
    },
  ];
};
