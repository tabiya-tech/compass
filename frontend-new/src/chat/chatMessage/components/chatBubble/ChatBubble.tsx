import { ConversationMessageSender } from "src/chat/ChatService/ChatService.types";
import { Box, Typography, styled, alpha, Divider, useTheme } from "@mui/material";

export interface ChatBubbleProps {
  message: string,
  sender: ConversationMessageSender,
  footer?: React.ReactNode
}

const uniqueId = "6e685eeb-2b54-432a-8b66-8a81633b3981";

export const DATA_TEST_ID = {
  CHAT_MESSAGE_BUBBLE_CONTAINER: `chat-message-bubble-container-${uniqueId}`,
  CHAT_MESSAGE_BUBBLE_MESSAGE_TEXT: `chat-message-bubble-message-text-${uniqueId}`,
  CHAT_MESSAGE_BUBBLE_MESSAGE_DIVIDER: `chat-message-bubble-message-divider-${uniqueId}`,
  CHAT_MESSAGE_BUBBLE_MESSAGE_FOOTER_CONTAINER: `chat-message-bubble-message-footer-container-${uniqueId}`,
}

const MessageBubble = styled(Box)<{ origin: ConversationMessageSender }>(({ theme, origin }) => ({
  maxWidth: "80%",
  minWidth: "30%",
  variants: "outlined",
  wordWrap: "break-word",
  padding: theme.fixedSpacing(theme.tabiyaSpacing.sm),
  border: origin === ConversationMessageSender.USER ? `2px solid ${theme.palette.primary.light}` : "none",
  borderRadius: origin === ConversationMessageSender.USER ? "12px 0px 12px 12px" : "12px 12px 12px 0px",
  backgroundColor:
    origin === ConversationMessageSender.USER ? alpha(theme.palette.primary.light, 0.16) : theme.palette.grey[100],
  color: origin === ConversationMessageSender.USER ? theme.palette.primary.contrastText : theme.palette.text.primary,
  position: "relative",
  alignSelf: origin === ConversationMessageSender.USER ? "flex-end" : "flex-start",
  display: "flex",
  flexDirection: "column",
}));

const ChatBubble: React.FC<ChatBubbleProps> = ({ message, sender, footer }) => {
  const theme = useTheme();
  return (
    <MessageBubble origin={sender} data-testid={DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_CONTAINER}>
      <Typography whiteSpace="pre-line" data-testid={DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_MESSAGE_TEXT}>{message}</Typography>
      {
        footer && (
          <>
            <Divider
              color={theme.palette.grey[100]}
              sx={{ marginY: theme.spacing(1) }}
              data-testid={DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_MESSAGE_DIVIDER}
            />
            <Box data-testid={DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_MESSAGE_FOOTER_CONTAINER}>
              {footer}
            </Box>
          </>
        )
      }
    </MessageBubble>
  )
}

export default ChatBubble;