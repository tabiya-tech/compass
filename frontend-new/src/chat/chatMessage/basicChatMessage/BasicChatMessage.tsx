import React from "react";
import { Box, styled } from "@mui/material";
import { IChatMessage } from "src/chat/Chat.types";
import { ConversationMessageSender } from "src/chat/ChatService/ChatService.types";
import ChatMessageFooter from "src/chat/chatMessage/components/chatMessageFooter/ChatMessageFooter";
import ChatBubble from "src/chat/chatMessage/components/chatBubble/ChatBubble";

const uniqueId = "2fbaf2ef-9eab-485a-bd28-b4a164e18b06";

export const DATA_TEST_ID = {
  CHAT_MESSAGE_CONTAINER: `chat-message-container-${uniqueId}`,
};

export const MessageContainer = styled(Box)<{ origin: ConversationMessageSender }>(({ theme, origin }) => ({
  display: "flex",
  flexDirection: "column",
  alignItems: origin === ConversationMessageSender.USER ? "flex-end" : "flex-start",
  marginBottom: theme.spacing(theme.tabiyaSpacing.sm),
  width: "100%",
}));

type BasicChatMessageProps = {
  chatMessage: IChatMessage;
};

const BasicChatMessage: React.FC<BasicChatMessageProps> = ({ chatMessage }) => {
  return (
    <MessageContainer origin={chatMessage.sender} data-testid={DATA_TEST_ID.CHAT_MESSAGE_CONTAINER}>
      <ChatBubble message={chatMessage.message} sender={chatMessage.sender} />
      <ChatMessageFooter sentAt={chatMessage.sent_at} />
    </MessageContainer>
  );
};

export default BasicChatMessage;
