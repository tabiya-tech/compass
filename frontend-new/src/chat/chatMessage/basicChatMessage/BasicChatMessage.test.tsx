// mute the console
import "src/_test_utilities/consoleMock";

import BasicChatMessage, { DATA_TEST_ID } from "./BasicChatMessage";
import ChatMessageFooter, { DATA_TEST_ID as CHAT_MESSAGE_FOOTER_DATA_TEST_ID } from "src/chat/chatMessage/components/chatMessageFooter/ChatMessageFooter";
import ChatBubble, { DATA_TEST_ID  as CHAT_BUBBLE_DATA_TEST_ID } from "src/chat/chatMessage/components/chatBubble/ChatBubble";
import { render, screen } from "src/_test_utilities/test-utils";
import { ConversationMessageSender } from "src/chat/ChatService/ChatService.types";
import { nanoid } from "nanoid";
import { ChatMessageType, IChatMessage } from "src/chat/Chat.types";

jest.mock("src/chat/chatMessage/components/chatMessageFooter/ChatMessageFooter", () => {
  const originalModule = jest.requireActual("src/chat/chatMessage/components/chatMessageFooter/ChatMessageFooter");
  return {
    __esModule: true,
    ...originalModule,
    default: jest.fn(() => <div data-testid={originalModule.DATA_TEST_ID.CHAT_MESSAGE_TIMESTAMP}></div>),
  }
})

jest.mock("src/chat/chatMessage/components/chatBubble/ChatBubble", () => {
  const originalModule = jest.requireActual("src/chat/chatMessage/components/chatBubble/ChatBubble");
  return {
    __esModule: true,
    ...originalModule,
    default: jest.fn(() => <div data-testid={originalModule.DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_CONTAINER}></div>),
  }
})

describe("render tests", () => {
  beforeAll(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date(2020, 3, 1));
  });

  afterAll(() => {
    jest.useRealTimers();
  });

  test("should render the Chat message with a timestamp", () => {
    // GIVEN a basic chat message sent at a given time
    const givenDate = new Date(2024, 6, 25).toISOString();
    const givenMessage: IChatMessage = {
      message_id: nanoid(),
      sender: ConversationMessageSender.COMPASS,
      message: "Hello, I'm Compass",
      sent_at: givenDate,
      type: ChatMessageType.BASIC_CHAT, // This component is designed for use with the BASIC chat type
    };

    // WHEN the chat message is rendered
    render(<BasicChatMessage chatMessage={givenMessage}/>);

    // THEN expect the message container to be visible
    expect(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_CONTAINER)).toBeInTheDocument();
    // AND expect the message bubble to be visible
    expect(screen.getByTestId(CHAT_BUBBLE_DATA_TEST_ID.CHAT_MESSAGE_BUBBLE_CONTAINER)).toBeInTheDocument();
    // AND expect the timestamp to be visible
    expect(screen.getByTestId(CHAT_MESSAGE_FOOTER_DATA_TEST_ID.CHAT_MESSAGE_TIMESTAMP)).toBeInTheDocument();

    // AND the correct date to have been displayed
    expect(ChatMessageFooter).toHaveBeenNthCalledWith(
      1,
      {
        sentAt: givenDate
      },
      {}
    )
    // AND expect the Chat bubble to have been rendered with the expected message
    expect(ChatBubble).toHaveBeenNthCalledWith(
      1,
      {
        message: givenMessage.message,
        sender: givenMessage.sender
      },
      {}
    )

    // AND expect the component to match the snapshot
    expect(screen.getByTestId(DATA_TEST_ID.CHAT_MESSAGE_CONTAINER)).toMatchSnapshot();
  });
});
