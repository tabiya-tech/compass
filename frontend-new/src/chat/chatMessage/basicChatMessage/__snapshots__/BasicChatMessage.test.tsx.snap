// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render tests should render the Chat message with a timestamp 1`] = `
<div
  class="MuiBox-root css-1c8zvho"
  data-testid="chat-message-container-2fbaf2ef-9eab-485a-bd28-b4a164e18b06"
  origin="COMPASS"
>
  <div
    data-testid="chat-message-bubble-container-6e685eeb-2b54-432a-8b66-8a81633b3981"
  />
  <div
    data-testid="chat-message-sent_at-7772f20a-9d0c-4072-b24f-97eca2f43d7b"
  />
</div>
`;
